diff --git a/src/shims/yield.c b/src/shims/yield.c
15e9f55..bea000c
--- a/swift-corelibs-libdispatch/src/shims/yield.c	2021-05-30 09:02:50.773811517 +0000
+++ b/swift-corelibs-libdispatch/src/shims/yield.c	2021-05-30 09:03:18.773260339 +0000
@@ -36,17 +36,6 @@
 _dispatch_wait_for_enqueuer(void **ptr)
 {
 #if !DISPATCH_HW_CONFIG_UP
-#if defined(__arm__) || defined(__arm64__)
-	int spins = DISPATCH_WAIT_SPINS_WFE;
-	void *value;
-	while (unlikely(spins-- > 0)) {
-		if (likely(value = __builtin_arm_ldrex(ptr))) {
-			__builtin_arm_clrex();
-			return value;
-		}
-		__builtin_arm_wfe();
-	}
-#else
 	int spins = DISPATCH_WAIT_SPINS;
 	void *value;
 	while (unlikely(spins-- > 0)) {
@@ -55,7 +44,6 @@
 		}
 		dispatch_hardware_pause();
 	}
-#endif
 #endif // DISPATCH_HW_CONFIG_UP
 	return __DISPATCH_WAIT_FOR_ENQUEUER__(ptr);
 }
