diff --git a/stdlib/public/SwiftShims/LibcShims.h b/stdlib/public/SwiftShims/LibcShims.h
index e3246b62fc..64763fe923 100644
--- a/stdlib/public/SwiftShims/LibcShims.h
+++ b/stdlib/public/SwiftShims/LibcShims.h
@@ -35,6 +35,8 @@ namespace swift { extern "C" {
 // the current platform in the runtime code.
 #if defined(__linux__) && defined (__arm__)
 typedef           int __swift_ssize_t;
+#elif defined(__linux__) && defined(__i386__)
+typedef           int __swift_ssize_t;
 #elif defined(_WIN32)
 #if defined(_M_ARM) || defined(_M_IX86)
 typedef           int __swift_ssize_t;
