diff --git a/include/swift/SIL/SILInstruction.h b/include/swift/SIL/SILInstruction.h
index 34eef2d..992f025 100644
--- a/include/swift/SIL/SILInstruction.h
+++ b/include/swift/SIL/SILInstruction.h
@@ -2797,12 +2797,12 @@ public:
   /// Return the type parameters to the builtin.
   SubstitutionList getSubstitutions() const {
     return {getTrailingObjects<Substitution>(),
-            SILInstruction::Bits.BuiltinInst.NumSubstitutions};
+            static_cast<size_t>(SILInstruction::Bits.BuiltinInst.NumSubstitutions)};
   }
   /// Return the type parameters to the builtin.
   MutableArrayRef<Substitution> getSubstitutions() {
     return {getTrailingObjects<Substitution>(),
-            SILInstruction::Bits.BuiltinInst.NumSubstitutions};
+            static_cast<size_t>(SILInstruction::Bits.BuiltinInst.NumSubstitutions)};
   }
   
   /// The arguments to the builtin.
