diff --git a/include/swift/Basic/Malloc.h b/include/swift/Basic/Malloc.h
index 4f937da..25c364e 100644
--- a/include/swift/Basic/Malloc.h
+++ b/include/swift/Basic/Malloc.h
@@ -35,7 +35,10 @@ inline void *AlignedAlloc(size_t size, size_t align) {
     align = sizeof(void*);
   
   void *r;
-#if defined(_WIN32)
+#if (defined(__linux__) && defined(__arm__))
+  r = calloc(1,size);
+  assert(r && "malloc failed");
+#elif defined(_WIN32)
   r = _aligned_malloc(size, align);
   assert(r && "_aligned_malloc failed");
 #else
diff --git a/stdlib/public/SwiftShims/RefCount.h b/stdlib/public/SwiftShims/RefCount.h
index e1cd7aa..9a55134 100644
--- a/stdlib/public/SwiftShims/RefCount.h
+++ b/stdlib/public/SwiftShims/RefCount.h
@@ -1302,7 +1302,7 @@ class HeapObjectSideTableEntry {
 
   public:
   HeapObjectSideTableEntry(HeapObject *newObject)
-    : object(newObject), refCounts()
+    : object(newObject), refCounts(RefCounts<SideTableRefCountBits>::Initialized)
   { }
 
 #pragma clang diagnostic push
