diff --git a/Sources/TSCUtility/Triple.swift b/Sources/TSCUtility/Triple.swift
index 5d5b45c6413..18c1ac83034 100644
--- a/Sources/TSCUtility/Triple.swift	2021-05-31 14:15:54.605849942 +0000
+++ b/Sources/TSCUtility/Triple.swift	2021-05-31 14:16:04.475652813 +0000
@@ -40,6 +40,7 @@
         case s390x
         case aarch64
         case armv7
+	case armv6
         case arm
         case arm64
         case arm64e
