diff --git a/cmake/modules/AddSwiftUnittests.cmake b/cmake/modules/AddSwiftUnittests.cmake
index f631429a536..9c9047cb3d7 100644
--- a/cmake/modules/AddSwiftUnittests.cmake
+++ b/cmake/modules/AddSwiftUnittests.cmake
@@ -56,6 +56,9 @@ function(add_swift_unittest test_dirname)
     if(CMAKE_SYSTEM_PROCESSOR MATCHES "x86_64|AMD64")
       target_compile_options(${test_dirname} PRIVATE
         -march=core2)
+    elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "armv6|armv7|i686")
+      set_property(TARGET "${test_dirname}" APPEND PROPERTY LINK_LIBRARIES
+        "atomic")
     endif()
   elseif("${SWIFT_HOST_VARIANT}" STREQUAL "windows")
     target_compile_definitions("${test_dirname}" PRIVATE
