diff --git a/lib/BuildSystem/BuildSystemFrontend.cpp b/lib/BuildSystem/BuildSystemFrontend.cpp
index 70740a5..25386c5 100644
--- a/lib/BuildSystem/BuildSystemFrontend.cpp
+++ b/lib/BuildSystem/BuildSystemFrontend.cpp
@@ -462,7 +462,10 @@ BuildSystemFrontendDelegate::createExecutionQueue() {
       numLanes = numCPUs;
     }
   }
-
+
+  numLanes = 1;
+  printf("Setting number of lanes to 1!\n");
+
   return std::unique_ptr<ExecutionQueue>(
       createLaneBasedExecutionQueue(impl->executionQueueDelegate, numLanes,
                                     impl->invocation.schedulerAlgorithm,
