diff --git a/CoreFoundation/Base.subproj/CFRuntime.h b/CoreFoundation/Base.subproj/CFRuntime.h
index d7aa26c..0f02a3c 100644
--- a/CoreFoundation/Base.subproj/CFRuntime.h
+++ b/CoreFoundation/Base.subproj/CFRuntime.h
@@ -196,10 +196,18 @@ typedef struct __CFRuntimeBase {
     uint32_t _swift_strong_rc;
     uint32_t _swift_weak_rc;
     // This is for CF's use, and must match _NSCFType layout
+#if __LP64__
     _Atomic(uint64_t) _cfinfoa;
+#else
+    _Atomic(uint32_t) _cfinfoa;
+#endif
 } CFRuntimeBase;
 
+#if __LP64__
 #define INIT_CFRUNTIME_BASE(...) {0, _CF_CONSTANT_OBJECT_STRONG_RC, 0, 0x0000000000000080ULL}
+#else
+#define INIT_CFRUNTIME_BASE(...) {0, _CF_CONSTANT_OBJECT_STRONG_RC, 0, 0x00000080UL}
+#endif
 
 #else
 
