diff --git a/CoreFoundation/Base.subproj/CFSortFunctions.c b/CoreFoundation/Base.subproj/CFSortFunctions.c
index 05bd43e..381f599 100644
--- a/CoreFoundation/Base.subproj/CFSortFunctions.c
+++ b/CoreFoundation/Base.subproj/CFSortFunctions.c
@@ -53,6 +53,9 @@ enum {
 #define __checkint_int64_mul(x, y, err)         (x * y)
 #define __checkint_uint64_add(x, y, err)        (x + y)
 
+#define __checkint_int32_mul(x,y,err)           (x * y)
+#define __checkint_uint32_add(x,y,err)          (x + y)
+
 #endif
 
 enum {
diff --git a/CoreFoundation/RunLoop.subproj/CFRunLoop.c b/CoreFoundation/RunLoop.subproj/CFRunLoop.c
index da2e573..df5cf50 100644
--- a/CoreFoundation/RunLoop.subproj/CFRunLoop.c
+++ b/CoreFoundation/RunLoop.subproj/CFRunLoop.c
@@ -47,6 +47,8 @@ CF_INLINE uint64_t check_uint64_add(uint64_t x, uint64_t y, int32_t* err) {
 #if __HAS_DISPATCH__ && __has_include(<dispatch/private.h>)
 #include <dispatch/private.h>
 #elif __HAS_DISPATCH__
+typedef int dispatch_runloop_handle_t;
+typedef uint32_t mach_port_t;
 extern dispatch_queue_t _dispatch_runloop_root_queue_create_4CF(const char *_Nullable label, unsigned long flags);
 extern mach_port_t _dispatch_runloop_root_queue_get_port_4CF(dispatch_queue_t queue);
 extern void _dispatch_source_set_runloop_timer_4CF(dispatch_source_t source, dispatch_time_t start, uint64_t interval, uint64_t leeway);
