diff --git a/Sources/Build/Triple.swift b/Sources/Build/Triple.swift
index 1ce1556..cba6776 100644
--- a/Sources/Build/Triple.swift
+++ b/Sources/Build/Triple.swift
@@ -34,6 +34,7 @@ public struct Triple {
         case x86_64
         case armv7
         case s390x
+        case arm
     }

     public enum Vendor: String {
@@ -106,11 +107,14 @@ public struct Triple {
     public static let macOS = try! Triple("x86_64-apple-macosx10.10")
     public static let linux = try! Triple("x86_64-unknown-linux")
     public static let android = try! Triple("armv7-unknown-linux-androideabi")
+    public static let armLinux = try! Triple("armv7-unknown-linux-gnueabihf")

   #if os(macOS)
     public static let hostTriple: Triple = .macOS
   #elseif os(Linux) && arch(s390x)
     public static let hostTriple: Triple = try! Triple("s390x-unknown-linux")
+  #elseif os(Linux) && arch(arm)
+    public static let hostTriple: Triple = .armLinux
   #else
     public static let hostTriple: Triple = .linux
   #endif
diff --git a/Utilities/bootstrap b/Utilities/bootstrap
index 784a232..08203b9 100755
--- a/Utilities/bootstrap
+++ b/Utilities/bootstrap
@@ -893,6 +893,8 @@ def main():
         build_target = "x86_64-apple-macosx10.10"
     elif platform.processor() == 's390x':
         build_target = "s390x-unknown-linux"
+    elif platform.system() == 'Linux' and platform.machine().startswith("arm"):
+        build_target = 'armv7-unknown-linux-gnueabihf'
     else:
         build_target = 'x86_64-unknown-linux'

