diff --git a/Sources/Build/Triple.swift b/Sources/Build/Triple.swift
index 1ce1556..c0cc760 100644
--- a/Sources/Build/Triple.swift
+++ b/Sources/Build/Triple.swift
@@ -33,6 +33,7 @@ public struct Triple {
     public enum Arch: String {
         case x86_64
         case armv7
+	case aarch64
         case s390x
     }
 
@@ -109,6 +110,8 @@ public struct Triple {
 
   #if os(macOS)
     public static let hostTriple: Triple = .macOS
+  #elseif os(Linux) && arch(arm64)
+    public static let hostTriple: Triple = try! Triple("aarch64-unknown-linux")
   #elseif os(Linux) && arch(s390x)
     public static let hostTriple: Triple = try! Triple("s390x-unknown-linux")
   #else
diff --git a/Utilities/bootstrap b/Utilities/bootstrap
index 784a232..f5a245c 100755
--- a/Utilities/bootstrap
+++ b/Utilities/bootstrap
@@ -891,6 +891,8 @@ def main():
     # Compute the build paths.
     if platform.system() == 'Darwin':
         build_target = "x86_64-apple-macosx10.10"
+    elif platform.system() == 'Linux' and platform.processor() == 'aarch64':
+        build_target = "aarch64-unknown-linux"
     elif platform.processor() == 's390x':
         build_target = "s390x-unknown-linux"
     else:
