diff --git a/Sources/Build/BuildPlan.swift b/Sources/Build/BuildPlan.swift
index e0ef02a..2b521c9 100644
--- a/Sources/Build/BuildPlan.swift
+++ b/Sources/Build/BuildPlan.swift
@@ -141,6 +141,9 @@ public struct BuildParameters {
     /// `.swiftmodule`s.
     public let enableParseableModuleInterfaces: Bool
 
+    /// Maximum number of threads to use for compilation (0 means use number of CPUs)
+    public let numThreads: UInt32
+
     /// Checks if stdout stream is tty.
     fileprivate let isTTY: Bool = {
         guard let stream = stdoutStream.stream as? LocalFileOutputByteStream else {
@@ -169,7 +172,8 @@ public struct BuildParameters {
         sanitizers: EnabledSanitizers = EnabledSanitizers(),
         enableCodeCoverage: Bool = false,
         indexStoreMode: IndexStoreMode = .auto,
-        enableParseableModuleInterfaces: Bool = false
+        enableParseableModuleInterfaces: Bool = false,
+        numThreads: UInt32 = 0
     ) {
         self.dataPath = dataPath
         self.configuration = configuration
@@ -182,7 +187,8 @@ public struct BuildParameters {
         self.sanitizers = sanitizers
         self.enableCodeCoverage = enableCodeCoverage
         self.indexStoreMode = indexStoreMode
-        self.enableParseableModuleInterfaces = enableParseableModuleInterfaces
+        self.enableParseableModuleInterfaces = enableParseableModuleInterfaces
+	 self.numThreads = numThreads
     }
 
     /// Returns the compiler arguments for the index store, if enabled.
diff --git a/Sources/Build/ToolProtocol.swift b/Sources/Build/ToolProtocol.swift
index 07ad19b..43775e2 100644
--- a/Sources/Build/ToolProtocol.swift
+++ b/Sources/Build/ToolProtocol.swift
@@ -142,6 +142,6 @@ struct SwiftCompilerTool: ToolProtocol {
         stream <<< "    enable-whole-module-optimization: "
             <<< Format.asJSON(target.buildParameters.configuration == .release) <<< "\n"
         stream <<< "    num-threads: "
-            <<< Format.asJSON("\(SwiftCompilerTool.numThreads)") <<< "\n"
+            <<< Format.asJSON("\(target.buildParameters.numThreads)") <<< "\n"
     }
 }
diff --git a/Sources/Commands/Options.swift b/Sources/Commands/Options.swift
index ad69c22..80db177 100644
--- a/Sources/Commands/Options.swift
+++ b/Sources/Commands/Options.swift
@@ -74,5 +74,8 @@ public class ToolOptions {
     /// Whether to enable generation of `.swiftinterface`s alongside `.swiftmodule`s.
     public var shouldEnableParseableModuleInterfaces = false
 
+    /// Limit the number of parallel instances of the swift compiler
+    public var numThreads = UInt32(0)
+
     public required init() {}
 }
diff --git a/Sources/Commands/SwiftTool.swift b/Sources/Commands/SwiftTool.swift
index 1386444..b1fa591 100644
--- a/Sources/Commands/SwiftTool.swift
+++ b/Sources/Commands/SwiftTool.swift
@@ -376,6 +376,11 @@ public class SwiftTool<Options: ToolOptions> {
             option: parser.add(option: "--enable-parseable-module-interfaces", kind: Bool.self),
             to: { $0.shouldEnableParseableModuleInterfaces = $1 })
 
+        binder.bind(
+            option: parser.add(option: "--numThreads", shortName: "-j", kind: Int.self,
+                usage: "Limit the number of parallel instances of the swift compiler (0 for unlimited)"),
+            to: { $0.numThreads = UInt32($1) })
+
         // Let subclasses bind arguments.
         type(of: self).defineArguments(parser: parser, binder: binder)
 
@@ -603,7 +608,7 @@ public class SwiftTool<Options: ToolOptions> {
         guard let llbuildTargetName = try computeLLBuildTargetName(for: subset, buildParameters: parameters) else {
             return
         }
-        try runLLBuild(manifest: parameters.llbuildManifest, llbuildTarget: llbuildTargetName)
+        try runLLBuild(manifest: parameters.llbuildManifest, llbuildTarget: llbuildTargetName, numThreads: parameters.numThreads)
     }
 
     /// Build a subset of products and targets using swift-build-tool.
@@ -619,7 +624,8 @@ public class SwiftTool<Options: ToolOptions> {
         try llbuild.generateManifest(at: yaml)
 
         // Run llbuild.
-        try runLLBuild(manifest: yaml, llbuildTarget: llbuildTargetName)
+        try runLLBuild(manifest: yaml, llbuildTarget: llbuildTargetName,
+		numThreads: plan.buildParameters.numThreads)
 
         // Create backwards-compatibilty symlink to old build path.
         let oldBuildPath = buildPath.appending(component: options.configuration.dirname)
@@ -629,17 +635,18 @@ public class SwiftTool<Options: ToolOptions> {
         try createSymlink(oldBuildPath, pointingAt: plan.buildParameters.buildPath, relative: true)
     }
 
-    func runLLBuild(manifest: AbsolutePath, llbuildTarget: String) throws {
+    func runLLBuild(manifest: AbsolutePath, llbuildTarget: String, numThreads: UInt32) throws {
         assert(localFileSystem.isFile(manifest), "llbuild manifest not present: \(manifest)")
         if options.shouldEnableLLBuildLibrary {
-            try runLLBuildAsLibrary(manifest: manifest, llbuildTarget: llbuildTarget)
+            try runLLBuildAsLibrary(manifest: manifest, llbuildTarget: llbuildTarget, numThreads: numThreads)
         } else {
-            try runLLBuildAsExecutable(manifest: manifest, llbuildTarget: llbuildTarget)
+            try runLLBuildAsExecutable(manifest: manifest, llbuildTarget: llbuildTarget, numThreads: numThreads)
         }
     }
 
-    func runLLBuildAsLibrary(manifest: AbsolutePath, llbuildTarget: String) throws {
+    func runLLBuildAsLibrary(manifest: AbsolutePath, llbuildTarget: String, numThreads: UInt32) throws {
         let databasePath = buildPath.appending(component: "build.db").pathString
+        BuildSystem.setSchedulerLaneWidth(width: numThreads)
         let buildSystem = BuildSystem(buildFile: manifest.pathString, databaseFile: databasePath, delegate: buildDelegate)
         buildDelegate.isVerbose = verbosity != .concise
         buildDelegate.onCommmandFailure = { [weak buildSystem] in buildSystem?.cancel() }
@@ -649,7 +656,7 @@ public class SwiftTool<Options: ToolOptions> {
         guard success else { throw Diagnostics.fatalError }
     }
 
-    func runLLBuildAsExecutable(manifest: AbsolutePath, llbuildTarget: String) throws {
+    func runLLBuildAsExecutable(manifest: AbsolutePath, llbuildTarget: String, numThreads: UInt32) throws {
         // Create a temporary directory for the build process.
         let tempDirName = "org.swift.swiftpm.\(NSUserName())"
         let tempDir = try determineTempDirectory().appending(component: tempDirName)
@@ -676,6 +683,9 @@ public class SwiftTool<Options: ToolOptions> {
         if verbosity != .concise {
             args.append("-v")
         }
+        if numThreads != 0 {
+            args.append("-j(\(numThreads))")
+        }
 
         // Create the environment for llbuild.
         var env = Process.env
@@ -722,7 +732,8 @@ public class SwiftTool<Options: ToolOptions> {
                 sanitizers: options.sanitizers,
                 enableCodeCoverage: options.shouldEnableCodeCoverage,
                 indexStoreMode: options.indexStoreMode,
-                enableParseableModuleInterfaces: options.shouldEnableParseableModuleInterfaces
+                enableParseableModuleInterfaces: options.shouldEnableParseableModuleInterfaces,
+                numThreads: options.numThreads
             )
         })
     }()
