diff --git a/lib/Driver/ToolChains.cpp b/lib/Driver/ToolChains.cpp
index c581bb460d..d69ac0f597 100644
--- a/lib/Driver/ToolChains.cpp
+++ b/lib/Driver/ToolChains.cpp
@@ -1753,6 +1753,7 @@ std::string toolchains::GenericUnix::getDefaultLinker() const {
     // linker.
     return "gold";
   case llvm::Triple::x86_64:
+  case llvm::Triple::aarch64:
   case llvm::Triple::ppc64:
   case llvm::Triple::ppc64le:
   case llvm::Triple::systemz:
