diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index c9c55b5..fa07c45 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -207,6 +207,8 @@ add_custom_command(TARGET dispatch POST_BUILD
                    COMMAND cmake -E copy $<TARGET_FILE:dispatch> .libs
                    COMMENT "Copying libdispatch to .libs")
 
+get_swift_host_arch(SWIFT_HOST_ARCH)
+
 install(TARGETS
           dispatch
         DESTINATION
@@ -216,6 +218,6 @@ if(ENABLE_SWIFT)
             ${CMAKE_CURRENT_BINARY_DIR}/swift/Dispatch.swiftmodule
             ${CMAKE_CURRENT_BINARY_DIR}/swift/Dispatch.swiftdoc
           DESTINATION
-            "${INSTALL_TARGET_DIR}/${CMAKE_SYSTEM_PROCESSOR}")
+          "${INSTALL_TARGET_DIR}/${SWIFT_HOST_ARCH}")
 endif()
 
