diff --git a/configure.ac b/configure.ac
index 04ba663..d999449 100644
--- a/configure.ac
+++ b/configure.ac
@@ -147,6 +147,15 @@ AC_ARG_WITH([swift-toolchain],
    case $target_os in
       linux*)
            os_string="linux"
+           case $target_cpu in
+                   armv7l*)
+                       target_cpu="armv7"
+                       ;;
+                   armv6l*)
+                       target_cpu="armv6"
+                       ;;
+                       *)
+            esac
            ;;
          *)
         os_string=$target_os
