diff --git a/icu4c/source/config/mh-linux b/icu4c/source/config/mh-linux
index 53d6780d68..c375aeb96c 100644
--- a/icu/icu4c/source/config/mh-linux
+++ b/icu/icu4c/source/config/mh-linux
@@ -23,7 +23,8 @@ LD_RPATH= -Wl,-zorigin,-rpath,'$$'ORIGIN
 LD_RPATH_PRE = -Wl,-rpath,
 
 ## These are the library specific LDFLAGS
-LDFLAGSICUDT=-nodefaultlibs -nostdlib
+#LDFLAGSICUDT=-nodefaultlibs -nostdlib
+LDFLAGSICUDT=
 
 ## Compiler switch to embed a library name
 # The initial tab in the next line is to prevent icu-config from reading it.
